packages/block-library/src/gallery/edit.js

```diff
@@ -312,6 +314,14 @@ class GalleryEdit extends Component {
 		this.setAttributes( { images: updatedImages, sizeSlug } );
 	}
 
+	/**
+	 * Sets the images in our array without redifining reference.
```

Comment:
Why we can not redefine the reference? It seems like we could remove this.images and simply pass this.props.attributes.images, if that's not the case maybe we have an existing issue in the code?

Reply:
> if that's not the case maybe we have an existing issue in the code?

Yeah, I'l have to dig a little further to see where that issue is stemming from.  For whatever reason, when we go from 0 images to some images the array passed down in props.attributes.images is redefined, so the uploading images were holding onto an outdated reference (empty array) for the currentMedia of the gallery.  This was causing the bug you noted above when trying to add 2 images to the empty gallery.  It could be something as simple as `images = galleryImages || []` up the chain somewhere, passing us some fallback empty array before things have been defined.  But yeah, if thats the case it will probably be best to clean it up at the source instead.

Reply:
I'm not quite sure if there is anything else we can easily do about this one.  It looks like the issue is the default supplied for the image array:

https://github.com/WordPress/gutenberg/blob/0336a3417024537a58cab3114cfbb12a5adabc83/packages/block-library/src/gallery/deprecated.js#L22

So when the gallery is new and has no images, props.attributes.images returns the default value of [].  Once there are images on the gallery, the actual gallery image array is passed instead.  This changing of the array we are referencing is what is causing that issue, the simplest fix I can think of is asserting control over them inside of the gallery component like I did, but am open to other suggestions.  🤔

Reply:
I encountered a similar issue on mobile, with a stale reference to the collection, and used the `useRef` hook to work around it: https://github.com/WordPress/gutenberg/blob/c6a9ab6745c0a97d3b8b70d4c1b5f54d4dd7c979/packages/block-editor/src/components/media-placeholder/index.native.js#L48

Upon further reflection, perhaps it'd be more idiomatic to use a more functional interface for the `value` prop, something like `{ otherProps, getCurrentValue = () => [], } = props;` and then used like: `<MediaPlaceholder value={ () => images }`, which would defer resolution of the collection reference, eliminating staleness arising from the asynchronicity. Wdyt?

Reply:
Note: if we go that route, we would need a minor change in `value` prop (e.g. replacing `value={ this.props.attributes }` with `value={ () => this.props.attributes }`) for the following blocks: Video, Gallery, Image, Audio, and the MediaPlaceholder README would need to be updated to reflect the change.

Reply:
Thanks for taking a look @mkevins!  That is a very neat implementation of `useRef`!  I was excited to implement it here and then realized `useRef` can only be used within functional components while both our gallery and mediaPlaceholder are class based currently. 
 Additionally, I am unable to find a similar class centric solution using `createRef` that works for this 🤔 .  Setting `this.images` seems to be the class based solution I can get working, and it works very similarly with an array instead of object.

> perhaps it'd be more idiomatic to use a more functional interface... if we go that route, we would need a minor change in value prop for the following blocks: Video, Gallery, Image, Audio, and the MediaPlaceholder

That is a thought, but it would depend on all of those being functional components.  So we would need to do some refactoring across the board where necessary for this.  Additionally, the only reason I see a need arising here is because the Gallery displays multiple media items at a time which makes this uploading process a bit more complex if we want dynamic behavior throughout that process.  Blocks that only display 1 media item don't really have the need for something like this as far as I can tell?

Reply:
Hi @Addison-Stavlo :wave: I think my suggestion may have been misinterpretted / misunderstood, (I could have worded it better :sweat_smile: )

> That is a thought, but it would depend on all of those being functional components. So we would need to do some refactoring across the board where necessary for this.

To clarify, the `useRef` workaround I had used in the past is the functional-component equivalent of exactly your class-component solution (i.e. the `useRef` is analogous to using `this.images` instance property). What I meant for an "alternative" solution is that instead of passing the array reference as the `value` prop to MediaPlaceholder (which gets stale), we could instead pass a function called `getCurrentValue`, which returns the images array when invoked (not stale). This should be possible for both functional components and class-components.

<details>
<summary>So onFilesUpload would look something like this:</summary>

```jsx
	onFilesUpload( files ) {
		const {
			addToGallery,
			allowedTypes,
			mediaUpload,
			multiple,
			onError,
			onSelect,
			// value = [],
			getCurrentValue = () => [],
		} = this.props;
		let setMedia;
		if ( multiple ) {
			if ( addToGallery ) {
				// const currentValue = value;
				setMedia = ( newMedia ) => {
					// onSelect( currentValue.concat( newMedia ) );
					onSelect( [ ...getCurrentValue(), ...newMedia ] );
				};
			} else {
				setMedia = onSelect;
			}
		} else {
			setMedia = ( [ media ] ) => onSelect( media );
		}
		mediaUpload( {
			allowedTypes,
			filesList: files,
			onFileChange: setMedia,
			onError,
		} );
	}
```
</details>

<details>
<summary>And Gallery's edit.js mediaPlaceholder would look something like this:</summary>

```jsx
		const mediaPlaceholder = (
			<MediaPlaceholder
				addToGallery={ hasImagesWithId }
				isAppender={ hasImages }
				className={ className }
				disableMediaButtons={ hasImages && ! isSelected }
				icon={ ! hasImages && sharedIcon }
				labels={ {
					title: ! hasImages && __( 'Gallery' ),
					instructions: ! hasImages && PLACEHOLDER_TEXT,
				} }
				onSelect={ this.onSelectImages }
				accept="image/*"
				allowedTypes={ ALLOWED_MEDIA_TYPES }
				multiple
				// value={ hasImagesWithId ? images : undefined }
				getCurrentValue={ () => images }
				onError={ this.onUploadError }
				notices={ hasImages ? undefined : noticeUI }
				onFocus={ this.props.onFocus }
			/>
		);
```
</details>

Does that help clarify?

Reply:
Oh gotcha, thanks for clarifying!  I do really like how that works.

That works in this case given that `getValue` is defined with this.props.attributes.images explicitly and not the dereferenced version.  `getCurrentValue={ () => this.props.attributes.images }`.  Then in the setMedia function I can call the getValue function instead of using the `currentMedia` I have defined there.

I do like this solution you are proposing, but I wonder if there would be any potential problems converting things to work this way?  Would we need backwards compatibility to support any 3rd party blocks that have been created using the value so they wouldn't break when we change this to be using the getValue?  In which case we should probably keep this current solution as well to provide this desired functionality in that case too?  If thats the case it may not be worth going down that road? 🤔

Reply:
> Would we need backwards compatibility to support any 3rd party blocks that have been created using the value so they wouldn't break when we change this to be using the getValue

Hm.. :thinking: I think we can use `value` as well, if needed, to support backward compatibility with third party blocks that have already been implemented using the current interface. Maybe something like this: `{ otherProp, value = [], getCurrentValue = () => value }` (we can use the `value` parameter in the default value for the `getCurrentValue` parameter, as long as it is listed _after_ the value parameter. IMO, this should satisfy backward compatibility for any previous 3rd party implementations.

Reply:
Ah yeah, we could do something like that.  I think in this case it would need to return this.props.value as returning the dereferenced version will become stale, but either way.. that will work!  So we could (fairly easily) refactor things to work this way and support backwards compatibility as well.

That would mean we could limit that change/refactor to just the placeholder and gallery block, and all other blocks could still work as is.  Code wise it does seem like the best solution, but I wonder what others will think about the necessity of this change.  Would adding extra props and updating the docs with this other option be worthwhile considering its only for 1 block that we already have an alternative working solution for that does not require that?  🤷‍♀️ Either way,  I'l branch off this PR and create that solution later today so we can have both available for consideration.  

Thanks again for your help here!

Reply:
I think I'm missing something in the problem with arrays changing.

> when we go from 0 images to some images the array passed down in props.attributes.images is redefined

> I encountered a similar issue on mobile, with a stale reference to the collection, and used the useRef hook to work around it


I guess it is expected that a value changes, and from one invocation to another, and a different value (reference) is passed. For example, if a user of the component passes a random array (redefined) of images every second, the component should update and work after each change. Could the issue be that somewhere in the code, we are relying on value always being the same reference? It would mean that the issue would not be that the array is redefined but that some code is relying on having a constant reference.

Reply:
> Could the issue be that somewhere in the code, we are relying on value always being the same reference? It would mean that the issue would not be that the array is redefined but that some code is relying on having a constant reference.

@jorgefilipecosta You are correct.  In order to get the desired dynamic functionality I added code that required the value always having the same reference.  It sounds like @mkevins did something similar and got around that by using the `useRef` hook, while here I got around that by using `this.images` since we aren't in a functional component.    

However, the new solution we were discussing now changes that.  Instead of having code that requires the reference to not change, this has been refactored to use a `getValue` function instead which eliminates the need to do all that mess that was going on with `this.images`.  I have just updated this (See e8615b4).